app/src/main/java/ch/epfl/culturequest/ArtDescriptionDisplayActivity.java

```diff
@@ -58,24 +61,93 @@ protected void onCreate(Bundle savedInstanceState) {
 
+        // Set Art Name
         TextView artNameView = findViewById(R.id.artName);
         setTextOrFallback(artNameView, artDescription.getName(), "No name found");
 
+        // Set Artist
         TextView artistNameView = findViewById(R.id.artistName);
         setTextOrFallback(artistNameView, artDescription.getArtist(), "No artist found");
 
+        // Set Year
         TextView artYearView = findViewById(R.id.artYear);
         setTextOrFallback(artYearView, artDescription.getYear(), "No year found");
 
+        // Set Summary
         TextView artSummaryView = findViewById(R.id.artSummary);
         setTextOrFallback(artSummaryView, artDescription.getSummary(), "No description found");
 
+        // Set Score
         TextView artScoreView = findViewById(R.id.artScore);
-        artScoreView.setText(artDescription.getScore() != null ? artDescription.getScore().toString() : "50");
+        artScoreView.setText(artDescription.getScore() != null ? "+" + artDescription.getScore().toString() + " pts" : "+30 pts");
+
+        // Set Rarity Badge
+        ImageView rarityBadge = findViewById(R.id.rarity);
+        setRarityBadge(rarityBadge, artDescription.getScore());
+
+        // Set Country Badge
+        ImageView countryBadge = findViewById(R.id.countryBadge);
+        TextView countryName = findViewById(R.id.countryName);
+
+        setCountryBadge(countryBadge, countryName, artDescription.getCountry());
+
+        // Set City Badge
+        ImageView cityBadge = findViewById(R.id.cityBadge);
+        TextView cityName = findViewById(R.id.cityName);
+
+        setCityBadge(cityBadge, cityName, artDescription.getCity());
+
+        // Set Museum Badge
+        ImageView museumBadge = findViewById(R.id.museumBadge);
+        TextView museumName = findViewById(R.id.museumName);
+
+        setMuseumBadge(museumBadge, museumName, artDescription.getMuseum());
     }
 
     private void setTextOrFallback(TextView textView, String text, String fallbackText) {
         textView.setText(text != null ? text : fallbackText);
     }
 
+    private void setRarityBadge(ImageView rarityBadge, Integer score) {
+        if (score != null) {
+            rarityBadge.setImageResource(getRarityLevel(score).getRarenessIcon());
+            rarityBadge.setTag(getRarityLevel(score).name());
+        } else {
+            rarityBadge.setImageResource(getRarityLevel(30).getRarenessIcon());
+            rarityBadge.setTag(getRarityLevel(30).name());
+        }
+    }
+
+    private void setCountryBadge(ImageView countryBadge, TextView countryText, String country) {
+        if (country != null) {
+            countryBadge.setImageResource(ScanBadge.Country.fromString(country).getBadge());
+            countryText.setText(country);
+            countryBadge.setTag(ScanBadge.Country.fromString(country).name());
+        } else {
+            countryBadge.setVisibility(ImageView.GONE);
+            countryText.setVisibility(TextView.GONE);
+        }
+    }
```

Comment:
Couldn't you generalize these 3 functions ? Maybe by passing the ScanBadge.X with X = country, City or Museum as as argument as well?

Reply:
It would not make sense as we would still end up with big switch cases for hardcoding the translation or ressources. 

We may need later to look for a real library that finds the normalized name of a place, but for now this does the job.

Reply:
Yeah I guess I didn't see the underlying implications. I think you could still modularize the `else` block in a separate function, but the gain on that part would not be that interesting.
In that case you're good to go!